# MP2 Part 1 — retrieve commits (netid `hvo5`)

Retrieves every commit World of Code has for the 10 projects assigned to `hvo5` in `net2prj.csv`, then records GitHub stars, forks, and the latest commit date.

World of Code (worldofcode.org) is a research archive of open-source version-control data, queried here through its HTTP API. It is not a live mirror of GitHub. The assignment says WoC data ends around January 2025, but the version served by this API includes commits into 2026. Its data still lags behind GitHub, so GitHub shows later commits for most projects. WoC project histories also include commits that are not on the GitHub default branch.

Commit content is requested in chunks of 10 (the API batch size, and under the 50-commit cap) with a 1 second wait between chunks. Results are cached in `df_commits.csv` and `df_commit_data.csv`, so a rerun continues where the last one stopped.


In [ ]:
%pip install -q -U python-woc pandas matplotlib requests tqdm
# please clear the output of this cell in your notebook before checking it in


In [2]:
from pathlib import Path
import csv
import os
import re
import subprocess
import time

import pandas as pd
import requests
from woc.remote import WocMapsRemote

assert Path("net2prj.csv").exists(), "Run this notebook from the folder that contains net2prj.csv."

NETID = "hvo5"
CHUNK = 10
SLEEP_SECONDS = 1
WOC_URL = "https://worldofcode.org/api/"

woc = WocMapsRemote(base_url=WOC_URL)
print("WoC client ready")


WoC client ready


## Assigned projects

The project list comes from `net2prj.csv` (rows for netid `hvo5`). It is not hardcoded.


In [3]:
assigned = pd.read_csv("net2prj.csv")
assigned = assigned[assigned["netID"] == NETID].reset_index(drop=True)
print(f"{len(assigned)} projects assigned to {NETID}")
assigned


10 projects assigned to hvo5


,netID,WoC,GH
0,hvo5,juliaastro_astrolib.jl,https://github.com/JuliaAstro/AstroLib.jl
1,hvo5,cra-ros-pkg_robot_localization,https://github.com/cra-ros-pkg/robot_localization
2,hvo5,hmri-group_hmri-toolbox,https://github.com/hMRI-group/hMRI-toolbox
3,hvo5,tdda_tdda,https://github.com/tdda/tdda
4,hvo5,openspace_openspace,https://github.com/OpenSpace/OpenSpace
5,hvo5,photon-hdf5_phconvert,https://github.com/Photon-HDF5/phconvert
6,hvo5,nocions_letswave6,https://github.com/NOCIONS/letswave6
7,hvo5,flatsurf_flatsurf,https://github.com/flatsurf/flatsurf
8,hvo5,mano-b_microsec,https://github.com/MANO-B/MicroSEC
9,hvo5,ablab_spades,https://github.com/ablab/spades


## Commit SHA1s from World of Code

`p2c` maps a WoC project id to commit SHA1s. If an id fails, the notebook tries the id derived from the GitHub URL (`owner_repo`, lowercased), which is the conversion used in the example notebook.


In [4]:
def reset_client():
    global woc
    woc = WocMapsRemote(base_url=WOC_URL)
    return woc


def github_owner_repo(url):
    parts = url.rstrip("/").split("/")
    return parts[-2], parts[-1]


def alternate_woc_id(url):
    owner, repo = github_owner_repo(url)
    return f"{owner}/{repo}".lower().replace("/", "_", 2)


def call_with_retry(label, fn, attempts=8):
    delay = 5
    last = None
    for attempt in range(1, attempts + 1):
        try:
            return fn()
        except KeyError:
            raise
        except Exception as exc:
            last = exc
            print(f"{label}: attempt {attempt} failed ({type(exc).__name__}: {exc})")
            time.sleep(delay)
            delay = min(delay * 2, 120)
            if attempt % 3 == 0:
                reset_client()
    raise RuntimeError(f"{label} failed after {attempts} attempts: {last}")


def project_shas(woc_id):
    values = call_with_retry(woc_id, lambda: woc.get_values("p2c", woc_id))
    if values is None:
        return []
    return list(values)


sha_path = Path("df_commits.csv")
if sha_path.exists() and sha_path.stat().st_size > 0:
    df = pd.read_csv(sha_path)
    print(f"Loaded cached SHA1 list from {sha_path} ({len(df)} rows)")
else:
    rows = []
    lookup_notes = []
    for rec in assigned.itertuples(index=False):
        prj = rec.WoC
        try:
            commits = project_shas(prj)
        except KeyError as exc:
            alt = alternate_woc_id(rec.GH)
            print(f"{prj}: not found ({exc}). Trying {alt}")
            commits = project_shas(alt)
            lookup_notes.append(f"{rec.WoC} -> {alt}")
            prj = alt
        print(f"{prj}: {len(commits)} commits")
        if len(commits) == 0:
            print(f"WARNING: {prj} returned 0 commits")
        for sha in commits:
            rows.append({"project": prj, "sha1": sha})
    df = pd.DataFrame(rows).drop_duplicates()
    df.to_csv(sha_path, index=False)
    print("lookup changes:", lookup_notes or "none")

print("SHA1 counts")
df.groupby("project").size().rename("commits")


juliaastro_astrolib.jl: 660 commits
cra-ros-pkg_robot_localization: 2428 commits
hmri-group_hmri-toolbox: 1613 commits


tdda_tdda: 1121 commits


openspace_openspace: 25521 commits
photon-hdf5_phconvert: 751 commits
nocions_letswave6: 311 commits
flatsurf_flatsurf: 2080 commits
mano-b_microsec: 654 commits


ablab_spades: 18459 commits
lookup changes: none
SHA1 counts


project
ablab_spades                      18459
cra-ros-pkg_robot_localization     2428
flatsurf_flatsurf                  2080
hmri-group_hmri-toolbox            1613
juliaastro_astrolib.jl              660
mano-b_microsec                     654
nocions_letswave6                   311
openspace_openspace               25521
photon-hdf5_phconvert               751
tdda_tdda                          1121
Name: commits, dtype: int64

## Commit content

Each commit SHA1 is looked up in `commit.tch` in chunks of 10, with a 1 second wait between chunks. Failed chunks are retried with backoff. Rows already stored in `df_commit_data.csv` are skipped, so this cell can be rerun after an interruption.


In [5]:
COMMIT_FIELDS = [
    "project",
    "commit",
    "tree",
    "parent",
    "author",
    "author_time",
    "author_tz",
    "committer",
    "committer_time",
    "committer_tz",
    "message",
]


def unwrap_commit(value):
    if isinstance(value, list) and value and isinstance(value[0], (list, tuple)):
        return value[0]
    return value


def parse_record(sha, record):
    record = unwrap_commit(record)
    if not isinstance(record, (list, tuple)) or len(record) < 5:
        raise ValueError(f"unexpected commit shape for {sha}: {str(record)[:180]}")
    parents = record[1]
    if isinstance(parents, (list, tuple)):
        parent = " ".join(str(p) for p in parents)
    else:
        parent = "" if parents is None else str(parents)
    author, committer = record[2], record[3]
    message = record[4] if record[4] is not None else ""
    return {
        "commit": sha,
        "tree": record[0],
        "parent": parent,
        "author": author[0],
        "author_time": int(author[1]),
        "author_tz": author[2],
        "committer": committer[0],
        "committer_time": int(committer[1]),
        "committer_tz": committer[2],
        "message": message,
    }


def fetch_chunk(shas):
    delay = 5
    last = None
    for attempt in range(1, 13):
        try:
            res, err = woc.get_values_many("commit.tch", list(shas))
            res = {k: unwrap_commit(v) for k, v in (res or {}).items()}
            return res, err or {}
        except Exception as exc:
            last = exc
            print(f"chunk retry {attempt} ({type(exc).__name__}: {exc})")
            time.sleep(delay)
            delay = min(delay * 2, 180)
            if attempt % 3 == 0:
                reset_client()
    raise RuntimeError(f"chunk failed after retries: {last}")


checkpoint = Path("df_commit_data.csv")
done = set()
if checkpoint.exists() and checkpoint.stat().st_size > 0:
    previous = pd.read_csv(checkpoint, usecols=["project", "commit"])
    done = set(zip(previous["project"], previous["commit"]))
    print(f"Resuming: {len(done)} commits already in {checkpoint.name}")

pending = [
    (project, sha)
    for project, sha in zip(df["project"], df["sha1"])
    if (project, sha) not in done
]
print(f"{len(pending)} commits still to fetch, {df['project'].nunique()} projects, {len(df)} SHA rows")

write_header = not checkpoint.exists() or checkpoint.stat().st_size == 0
out = checkpoint.open("a", newline="", encoding="utf-8")
writer = csv.DictWriter(out, fieldnames=COMMIT_FIELDS, extrasaction="ignore")
if write_header:
    writer.writeheader()
    out.flush()

failed = []
saved_now = 0
started = time.time()


def flush_rows(rows):
    global saved_now
    if not rows:
        return
    writer.writerows(rows)
    out.flush()
    os.fsync(out.fileno())
    saved_now += len(rows)


buffer = []
try:
    for start in range(0, len(pending), CHUNK):
        chunk_pairs = pending[start:start + CHUNK]
        projects_for = {}
        for project, sha in chunk_pairs:
            projects_for.setdefault(sha, []).append(project)
        res, err = fetch_chunk(list(projects_for))
        if err:
            print(f"API errors for {len(err)} sha1s in this chunk: {list(err)[:3]}")
        for sha, projects in projects_for.items():
            if sha not in res or res[sha] is None:
                for project in projects:
                    failed.append((project, sha, str((err or {}).get(sha, "missing from response"))))
                continue
            try:
                parsed = parse_record(sha, res[sha])
            except Exception as exc:
                for project in projects:
                    failed.append((project, sha, str(exc)))
                continue
            for project in projects:
                buffer.append({"project": project, **parsed})
        if len(buffer) >= 200:
            flush_rows(buffer)
            buffer = []
            done_total = len(done) + saved_now
            elapsed = time.time() - started
            print(f"saved {done_total}/{len(df)} commits ({elapsed/60:.1f} min)")
        time.sleep(SLEEP_SECONDS)
    flush_rows(buffer)
    buffer = []
finally:
    out.close()

# Second pass for commits the API did not return.
if failed:
    print(f"retrying {len(failed)} commits that failed the first pass")
    still_failed = []
    out = checkpoint.open("a", newline="", encoding="utf-8")
    writer = csv.DictWriter(out, fieldnames=COMMIT_FIELDS, extrasaction="ignore")
    try:
        for start in range(0, len(failed), CHUNK):
            group = failed[start:start + CHUNK]
            projects_for = {}
            for project, sha, _reason in group:
                projects_for.setdefault(sha, []).append(project)
            res, err = fetch_chunk(list(projects_for))
            rows = []
            for sha, projects in projects_for.items():
                if sha not in res or res[sha] is None:
                    reason = str((err or {}).get(sha, "missing from response"))
                    for project in projects:
                        still_failed.append((project, sha, reason))
                    continue
                try:
                    parsed = parse_record(sha, res[sha])
                except Exception as exc:
                    for project in projects:
                        still_failed.append((project, sha, str(exc)))
                    continue
                for project in projects:
                    rows.append({"project": project, **parsed})
            flush_rows(rows)
            time.sleep(SLEEP_SECONDS)
    finally:
        out.close()
    failed = still_failed

print(f"fetch finished in {(time.time()-started)/60:.1f} min; unresolved commits: {len(failed)}")
if failed:
    print("unresolved sample:", failed[:5])
if len(failed) > max(20, int(0.01 * len(df))):
    raise RuntimeError(f"{len(failed)} commits are still missing; rerun this cell to resume")


53598 commits still to fetch, 10 projects, 53598 SHA rows


API errors for 1 sha1s in this chunk: ['2369e4df5b846622dd0a68e9ceffea658a97a9e6']


saved 209/53598 commits (0.3 min)


API errors for 1 sha1s in this chunk: ['5d674b56e40febecc4f47aacf3e84f27b5122dff']


API errors for 2 sha1s in this chunk: ['7834eb815c8205bf495530c0bf0cee627c2e7936', '7895504120d07d30488a21e033a11df812fdb271']


API errors for 1 sha1s in this chunk: ['864550f8f4fe20a52d9c1f486c66ad0fd267484c']


API errors for 1 sha1s in this chunk: ['9d97bd1e3ac9ce1273d2b948387c0a3f8409b306']


API errors for 1 sha1s in this chunk: ['9f1c76374d6de7d9db82559502acfa9c7ec95113']
saved 413/53598 commits (0.7 min)


API errors for 1 sha1s in this chunk: ['a228fea603ca3575be9aa24ecee5c08024da3b8d']


API errors for 1 sha1s in this chunk: ['cbf3dfe36f75b17ece25ed92d45440638c8477df']


saved 621/53598 commits (1.1 min)


API errors for 1 sha1s in this chunk: ['0597763a66c0470f1c6ae47d8dd7aeb19dfe08ab']


API errors for 1 sha1s in this chunk: ['09cfbe28f650db9d95224640645163ed57ed0718']


saved 829/53598 commits (1.5 min)


API errors for 1 sha1s in this chunk: ['24868d34ff5061a33effb97dc7c152df4f328f4e']


saved 1038/53598 commits (1.8 min)


API errors for 1 sha1s in this chunk: ['3211e48305a74c2000a862bdf36912760be8d3a2']


API errors for 1 sha1s in this chunk: ['33553f516b977e5b3e49c855b544e50fc3ed6b3c']


API errors for 1 sha1s in this chunk: ['352eb773e41ba4e42c92719b64c1de25806199b2']


saved 1245/53598 commits (2.2 min)


saved 1445/53598 commits (2.5 min)


API errors for 1 sha1s in this chunk: ['631b7aa663639e56e03258ec0af800c5665ed11c']


saved 1654/53598 commits (2.9 min)


API errors for 1 sha1s in this chunk: ['6f2af9b64e1761795d93160cb1a4abd7b47f8e29']


saved 1863/53598 commits (3.3 min)


saved 2063/53598 commits (3.6 min)


API errors for 1 sha1s in this chunk: ['9fdbcc89226602f9e82be4ceb2950d329ef25a3b']


saved 2272/53598 commits (4.0 min)


API errors for 1 sha1s in this chunk: ['b33dea459921599723cac7ee69966b3d917ff31c']


API errors for 1 sha1s in this chunk: ['bce4f30c0cbb1adcc2c123da10a3cd54555da590']
saved 2480/53598 commits (4.4 min)


API errors for 1 sha1s in this chunk: ['c0011c2e93ca6ec8e4759fe237c50fb3dd363673']


API errors for 1 sha1s in this chunk: ['d496da30a071f3d849a8f561245b5aba8b752521']
saved 2688/53598 commits (4.7 min)


saved 2888/53598 commits (5.1 min)


API errors for 1 sha1s in this chunk: ['ff7868ae536af9d8d936005f30809d1ef2b09c26']


saved 3097/53598 commits (5.4 min)


saved 3297/53598 commits (5.8 min)


API errors for 1 sha1s in this chunk: ['2ad986ae65ff13699ecfaa3f8925b60d8c787a11']


API errors for 1 sha1s in this chunk: ['3d7447ae3be7cad42aaf931c71eb1fd6edee8752']


API errors for 1 sha1s in this chunk: ['3e8ab40d05513d4a1bca1b7e8d7940c433244f50']


saved 3504/53598 commits (6.2 min)


saved 3704/53598 commits (6.5 min)


saved 3904/53598 commits (6.9 min)


API errors for 1 sha1s in this chunk: ['9841f07bf63ffb06ab3aa6e9bc670c6073052257']


saved 4113/53598 commits (7.2 min)


saved 4313/53598 commits (7.6 min)


saved 4513/53598 commits (7.9 min)


saved 4713/53598 commits (8.3 min)


API errors for 1 sha1s in this chunk: ['0d02dd84cd485a4d1b7d94c9cc240224de25f8d0']


API errors for 1 sha1s in this chunk: ['0ee1cf930f8aa583ffc59668b1ecdb2c8d425ba5']


API errors for 1 sha1s in this chunk: ['15e24ea64080553f25ae003d2ebcc08f26e5a6f9']


API errors for 1 sha1s in this chunk: ['1a33a7e643238899b76e03a6b45247b71169e7fa']


API errors for 1 sha1s in this chunk: ['254e0d31dd4ce98e85b2cf5ef773e967007f1a60']


API errors for 1 sha1s in this chunk: ['323c06f983165be5db69235f2d7782037f7b079b']


saved 4917/53598 commits (8.7 min)


API errors for 1 sha1s in this chunk: ['386b68abe301f59481a5d56bc636681018c4e206']


API errors for 1 sha1s in this chunk: ['51a7654935b20ad361ee437acb8f41c44a97a65a']


API errors for 1 sha1s in this chunk: ['5701d7fb3996d14622995f6cf802d3c1f3a08f17']


API errors for 1 sha1s in this chunk: ['6422861ccef4efd27d46d812f63079da7c64a9ac']


saved 5123/53598 commits (9.0 min)


API errors for 2 sha1s in this chunk: ['6a46c43042debec722c1ad914be37f2b2230d529', '6ae57a172bae08f8393d28995ba9cc6fea065900']


API errors for 1 sha1s in this chunk: ['7e39c53924a80f7cf382ee74ee21640ab96da475']


API errors for 1 sha1s in this chunk: ['80a9c2d538b8c2b5a4d6662276ce7ccb90ca3c7a']


API errors for 1 sha1s in this chunk: ['8804cc31d5ce7a11e93a72104fc0a1cd8511e3b1']


API errors for 1 sha1s in this chunk: ['8dfda91872e989f31c92724b0c28a69a3f8a43a3']


API errors for 1 sha1s in this chunk: ['962597dac6c4aef1a364c8605b40159f9b769f36']
saved 5326/53598 commits (9.4 min)


API errors for 1 sha1s in this chunk: ['9be59f48aa40c70bbfe1dfa9f9479da3e62bb6cc']


API errors for 1 sha1s in this chunk: ['ac73dcc3b05f83dc59fa315edd2331d841b8963e']


API errors for 1 sha1s in this chunk: ['b5396cce5fb2b07a85449c3a16cef077e391e5a3']


saved 5533/53598 commits (9.8 min)


API errors for 1 sha1s in this chunk: ['d4b20d9aa52653ed433bfc38bafe4443cca93aed']


API errors for 1 sha1s in this chunk: ['dc13e2aae5ffb311f6374dd992c5924868173ee7']


API errors for 1 sha1s in this chunk: ['e07c488d273a35bcac76c930050c4700a7579dbb']


API errors for 1 sha1s in this chunk: ['ebfb0547cf4ebf4a84cc0ee217fbe20e738ab445']


API errors for 1 sha1s in this chunk: ['ee88712db40cdbf18f230312e61a85e1df78954a']


saved 5738/53598 commits (10.2 min)


API errors for 1 sha1s in this chunk: ['f96a4e29da6746be6f85dab1f44d81d5f6111542']


API errors for 1 sha1s in this chunk: ['fce98d8d111e6fd635b00752577f7f11d9f23b07']


saved 5946/53598 commits (10.5 min)


API errors for 1 sha1s in this chunk: ['033bdd39560dab965a349862765cc6a5c1de2db7']


saved 6155/53598 commits (10.9 min)


API errors for 1 sha1s in this chunk: ['04812919c00b192e6cc1742c41000ec51cc042e9']


saved 6364/53598 commits (11.3 min)


saved 6564/53598 commits (11.6 min)


API errors for 1 sha1s in this chunk: ['0946de65aaa17beedf95c4ce5d5299568d4a67db']


saved 6773/53598 commits (12.0 min)


saved 6973/53598 commits (12.3 min)


saved 7173/53598 commits (12.7 min)


saved 7373/53598 commits (13.0 min)


saved 7573/53598 commits (13.4 min)


API errors for 1 sha1s in this chunk: ['1384b0a9393ebffd7c688453b2ad0410925535cd']


saved 7782/53598 commits (13.7 min)


saved 7982/53598 commits (14.1 min)


saved 8182/53598 commits (14.4 min)


saved 8382/53598 commits (14.8 min)


saved 8582/53598 commits (15.1 min)


saved 8782/53598 commits (15.5 min)


saved 8982/53598 commits (15.8 min)


saved 9182/53598 commits (16.2 min)


saved 9382/53598 commits (16.5 min)


API errors for 1 sha1s in this chunk: ['23effab57d3c4fc3550ef26ab6c5090f333b93fd']


saved 9591/53598 commits (16.9 min)


API errors for 1 sha1s in this chunk: ['260fa4f7c2d857cf02cfc682f7597aca2a9b664d']


saved 9800/53598 commits (17.3 min)


API errors for 1 sha1s in this chunk: ['298988d6584e95a8249c898a9dfcd4c42e65e08d']


saved 10009/53598 commits (17.6 min)


saved 10209/53598 commits (18.0 min)


saved 10409/53598 commits (18.3 min)


saved 10609/53598 commits (18.7 min)


saved 10809/53598 commits (19.0 min)


saved 11009/53598 commits (19.4 min)


API errors for 1 sha1s in this chunk: ['35370e643105fe9815fbf8592ed8df430b64b906']


API errors for 1 sha1s in this chunk: ['363456c14944fe1a9806a3fee976b2a13bc12f56']


API errors for 1 sha1s in this chunk: ['367d445c0f58bb40d43cb86aac6c73b1e78974a1']
saved 11216/53598 commits (19.7 min)


saved 11416/53598 commits (20.1 min)


API errors for 1 sha1s in this chunk: ['39d8aa9b3ecc09e603e04a5476846457f22ff52c']


saved 11625/53598 commits (20.5 min)


saved 11825/53598 commits (20.8 min)


saved 12025/53598 commits (21.2 min)


API errors for 1 sha1s in this chunk: ['3f7b926f6b105f14450e92fec284f1f844d56f2e']


saved 12234/53598 commits (21.5 min)


saved 12434/53598 commits (21.9 min)


saved 12634/53598 commits (22.2 min)


saved 12834/53598 commits (22.6 min)


saved 13034/53598 commits (22.9 min)


saved 13234/53598 commits (23.3 min)


saved 13434/53598 commits (23.6 min)


API errors for 1 sha1s in this chunk: ['4e59df3ca1f1e0574f974967d1956437fa707465']


saved 13643/53598 commits (24.0 min)


saved 13843/53598 commits (24.3 min)


saved 14043/53598 commits (24.7 min)


API errors for 1 sha1s in this chunk: ['53e7d195f11605c00e3ca4952b836df6e80bd1f3']


saved 14252/53598 commits (25.1 min)


saved 14452/53598 commits (25.4 min)


API errors for 1 sha1s in this chunk: ['58353af03f8a69155e4b1b67df40c580fba9cb2c']


saved 14661/53598 commits (25.8 min)


API errors for 1 sha1s in this chunk: ['594c80c8aa723a6777bc2cd2d85b0ff2c3f2b75d']


saved 14870/53598 commits (26.1 min)


API errors for 1 sha1s in this chunk: ['5c27d7f34f3b3abcaa87122e4bb772617c6f732f']


saved 15079/53598 commits (26.5 min)


saved 15279/53598 commits (26.9 min)


saved 15479/53598 commits (27.2 min)


saved 15679/53598 commits (27.6 min)


saved 15879/53598 commits (27.9 min)


saved 16079/53598 commits (28.3 min)


saved 16279/53598 commits (28.6 min)


saved 16479/53598 commits (29.0 min)


saved 16679/53598 commits (29.3 min)


saved 16879/53598 commits (29.7 min)


saved 17079/53598 commits (30.0 min)


saved 17279/53598 commits (30.4 min)


saved 17479/53598 commits (30.7 min)


saved 17679/53598 commits (31.1 min)


saved 17879/53598 commits (31.4 min)


saved 18079/53598 commits (31.8 min)


saved 18279/53598 commits (32.1 min)


API errors for 1 sha1s in this chunk: ['7e9bbb18fc876aed7a67f876c03c4662a16f2fba']


saved 18488/53598 commits (32.5 min)


saved 18688/53598 commits (32.8 min)


saved 18888/53598 commits (33.2 min)


saved 19088/53598 commits (33.5 min)


saved 19288/53598 commits (33.9 min)


saved 19488/53598 commits (34.2 min)


saved 19688/53598 commits (34.6 min)


saved 19888/53598 commits (34.9 min)


API errors for 1 sha1s in this chunk: ['908c39a496589227a46c77d4740b6492b5f1cf36']


saved 20097/53598 commits (35.3 min)


saved 20297/53598 commits (35.6 min)


API errors for 1 sha1s in this chunk: ['93ee636a5228226f3fd7001c6c6d0867c6d076f7']


saved 20506/53598 commits (36.0 min)


saved 20706/53598 commits (36.4 min)


saved 20906/53598 commits (36.7 min)


saved 21106/53598 commits (37.1 min)


API errors for 1 sha1s in this chunk: ['9b984c1afe4eefb05c30ff47e551892a6b5dc73a']


saved 21315/53598 commits (37.4 min)


saved 21515/53598 commits (37.8 min)


saved 21715/53598 commits (38.1 min)


saved 21915/53598 commits (38.5 min)


saved 22115/53598 commits (38.8 min)


API errors for 1 sha1s in this chunk: ['a673619e498d6184dcd4ea4e13da2127b212ae36']


saved 22324/53598 commits (39.2 min)


saved 22524/53598 commits (39.5 min)


saved 22724/53598 commits (39.9 min)


saved 22924/53598 commits (40.2 min)


saved 23124/53598 commits (40.6 min)


saved 23324/53598 commits (40.9 min)


saved 23524/53598 commits (41.3 min)


API errors for 1 sha1s in this chunk: ['b4a43166e9af66729ae59855e4ec13285aed9dbe']


saved 23733/53598 commits (41.6 min)


saved 23933/53598 commits (42.0 min)


API errors for 1 sha1s in this chunk: ['b883fd65d88542c3ffe617b398699a1f23ede604']


saved 24142/53598 commits (42.4 min)


saved 24342/53598 commits (42.7 min)


saved 24542/53598 commits (43.1 min)


saved 24742/53598 commits (43.4 min)


saved 24942/53598 commits (43.8 min)


saved 25142/53598 commits (44.1 min)


saved 25342/53598 commits (44.5 min)


API errors for 1 sha1s in this chunk: ['c5cdf5eba69afdc98394f8b7dc3467f1b1626937']


saved 25551/53598 commits (44.9 min)


saved 25751/53598 commits (45.2 min)


saved 25951/53598 commits (45.6 min)


saved 26151/53598 commits (45.9 min)


saved 26351/53598 commits (46.3 min)


saved 26551/53598 commits (46.6 min)


API errors for 1 sha1s in this chunk: ['d1e8fdd89cb8650232680e6887239a308531cfd0']


saved 26760/53598 commits (47.0 min)


saved 26960/53598 commits (47.3 min)


saved 27160/53598 commits (47.7 min)


API errors for 1 sha1s in this chunk: ['d6b33c15f9a8ec6814c5da7436fa7a860b768d21']


saved 27369/53598 commits (48.0 min)


saved 27569/53598 commits (48.4 min)


saved 27769/53598 commits (48.7 min)


saved 27969/53598 commits (49.1 min)


saved 28169/53598 commits (49.4 min)


API errors for 1 sha1s in this chunk: ['e14d94bd99cbe0534e18f89b187383b3f2b69c55']


saved 28378/53598 commits (49.8 min)


saved 28578/53598 commits (50.2 min)


saved 28778/53598 commits (50.5 min)


saved 28978/53598 commits (50.9 min)


saved 29178/53598 commits (51.2 min)


saved 29378/53598 commits (51.6 min)


saved 29578/53598 commits (51.9 min)


saved 29778/53598 commits (52.3 min)


saved 29978/53598 commits (52.6 min)


saved 30178/53598 commits (53.0 min)


saved 30378/53598 commits (53.3 min)


saved 30578/53598 commits (53.7 min)


saved 30778/53598 commits (54.0 min)


saved 30978/53598 commits (54.4 min)


saved 31178/53598 commits (54.7 min)


saved 31378/53598 commits (55.1 min)


saved 31578/53598 commits (55.4 min)


saved 31778/53598 commits (55.8 min)


saved 31978/53598 commits (56.1 min)


saved 32178/53598 commits (56.5 min)


saved 32378/53598 commits (56.8 min)


saved 32578/53598 commits (57.2 min)


saved 32778/53598 commits (57.5 min)


saved 32978/53598 commits (57.9 min)


saved 33178/53598 commits (58.2 min)


API errors for 1 sha1s in this chunk: ['7b49c8b71d45d1b1334f8cbc62ecdb0a671d51bc']


saved 33387/53598 commits (58.6 min)


saved 33587/53598 commits (58.9 min)


saved 33787/53598 commits (59.3 min)


saved 33987/53598 commits (59.6 min)


saved 34187/53598 commits (60.0 min)


saved 34387/53598 commits (60.3 min)


saved 34587/53598 commits (60.7 min)


saved 34787/53598 commits (61.0 min)


saved 34987/53598 commits (61.4 min)


saved 35187/53598 commits (61.7 min)


saved 35387/53598 commits (62.1 min)


saved 35587/53598 commits (62.4 min)


saved 35787/53598 commits (62.8 min)


saved 35987/53598 commits (63.1 min)


saved 36187/53598 commits (63.5 min)


saved 36387/53598 commits (63.8 min)


saved 36587/53598 commits (64.2 min)


saved 36787/53598 commits (64.5 min)


saved 36987/53598 commits (64.9 min)


saved 37187/53598 commits (65.2 min)


saved 37387/53598 commits (65.6 min)


saved 37587/53598 commits (65.9 min)


saved 37787/53598 commits (66.3 min)


saved 37987/53598 commits (66.6 min)


saved 38187/53598 commits (67.0 min)


saved 38387/53598 commits (67.3 min)


saved 38587/53598 commits (67.7 min)


saved 38787/53598 commits (68.0 min)


saved 38987/53598 commits (68.4 min)


saved 39187/53598 commits (68.7 min)


saved 39387/53598 commits (69.1 min)


saved 39587/53598 commits (69.4 min)


saved 39787/53598 commits (69.8 min)


saved 39987/53598 commits (70.1 min)


saved 40187/53598 commits (70.5 min)


saved 40387/53598 commits (70.8 min)


saved 40587/53598 commits (71.2 min)


saved 40787/53598 commits (71.5 min)


saved 40987/53598 commits (71.9 min)


saved 41187/53598 commits (72.2 min)


saved 41387/53598 commits (72.6 min)


saved 41587/53598 commits (72.9 min)


saved 41787/53598 commits (73.3 min)


saved 41987/53598 commits (73.6 min)


saved 42187/53598 commits (74.0 min)


saved 42387/53598 commits (74.3 min)


saved 42587/53598 commits (74.7 min)


saved 42787/53598 commits (75.0 min)


saved 42987/53598 commits (75.4 min)


saved 43187/53598 commits (75.7 min)


saved 43387/53598 commits (76.1 min)


saved 43587/53598 commits (76.4 min)


saved 43787/53598 commits (76.8 min)


saved 43987/53598 commits (77.1 min)


saved 44187/53598 commits (77.5 min)


saved 44387/53598 commits (77.8 min)


saved 44587/53598 commits (78.2 min)


saved 44787/53598 commits (78.5 min)


saved 44987/53598 commits (78.9 min)


saved 45187/53598 commits (79.2 min)


saved 45387/53598 commits (79.6 min)


saved 45587/53598 commits (79.9 min)


saved 45787/53598 commits (80.3 min)


saved 45987/53598 commits (80.6 min)


saved 46187/53598 commits (81.0 min)


saved 46387/53598 commits (81.3 min)


saved 46587/53598 commits (81.7 min)


saved 46787/53598 commits (82.0 min)


saved 46987/53598 commits (82.4 min)


saved 47187/53598 commits (82.7 min)


saved 47387/53598 commits (83.1 min)


saved 47587/53598 commits (83.4 min)


saved 47787/53598 commits (83.8 min)


saved 47987/53598 commits (84.1 min)


saved 48187/53598 commits (84.5 min)


saved 48387/53598 commits (84.8 min)


saved 48587/53598 commits (85.2 min)


saved 48787/53598 commits (85.5 min)


saved 48987/53598 commits (85.9 min)


saved 49187/53598 commits (86.2 min)


saved 49387/53598 commits (86.6 min)


saved 49587/53598 commits (86.9 min)


saved 49787/53598 commits (87.3 min)


saved 49987/53598 commits (87.6 min)


saved 50187/53598 commits (88.0 min)


saved 50387/53598 commits (88.3 min)


saved 50587/53598 commits (88.7 min)


saved 50787/53598 commits (89.0 min)


saved 50987/53598 commits (89.4 min)


saved 51187/53598 commits (89.7 min)


saved 51387/53598 commits (90.1 min)


saved 51587/53598 commits (90.4 min)


saved 51787/53598 commits (90.8 min)


saved 51987/53598 commits (91.1 min)


saved 52187/53598 commits (91.5 min)


saved 52387/53598 commits (91.8 min)


saved 52587/53598 commits (92.2 min)


saved 52787/53598 commits (92.5 min)


saved 52987/53598 commits (92.9 min)


saved 53187/53598 commits (93.2 min)


saved 53387/53598 commits (93.5 min)


retrying 83 commits that failed the first pass


fetch finished in 94.0 min; unresolved commits: 83
unresolved sample: [('juliaastro_astrolib.jl', '2369e4df5b846622dd0a68e9ceffea658a97a9e6', 'Key 2369e4df5b846622dd0a68e9ceffea658a97a9e6 not found in /da5_fast/All.sha1c/commit_35.tch'), ('juliaastro_astrolib.jl', '5d674b56e40febecc4f47aacf3e84f27b5122dff', 'Key 5d674b56e40febecc4f47aacf3e84f27b5122dff not found in /da5_fast/All.sha1c/commit_93.tch'), ('juliaastro_astrolib.jl', '7834eb815c8205bf495530c0bf0cee627c2e7936', 'Key 7834eb815c8205bf495530c0bf0cee627c2e7936 not found in /da5_fast/All.sha1c/commit_120.tch'), ('juliaastro_astrolib.jl', '7895504120d07d30488a21e033a11df812fdb271', 'Key 7895504120d07d30488a21e033a11df812fdb271 not found in /da5_fast/All.sha1c/commit_120.tch'), ('juliaastro_astrolib.jl', '864550f8f4fe20a52d9c1f486c66ad0fd267484c', 'Key 864550f8f4fe20a52d9c1f486c66ad0fd267484c not found in /da5_fast/All.sha1c/commit_6.tch')]


In [6]:
# One retrieved commit, so the saved output shows the WoC record shape.
sample = pd.read_csv("df_commit_data.csv", nrows=1)
sample.T


,0
project,juliaastro_astrolib.jl
commit,00430dee0d6de3e87fe945e90a84775a82c519a9
tree,42d155a84adb0d28bcca706622cf8d4f7921815b
parent,b2a83edd44cb6f15343165af27ef31c6bee6cfc9
author,Documenter.jl <documenter@juliadocs.github.io>
author_time,1622754964
author_tz,0
committer,Documenter.jl <documenter@juliadocs.github.io>
committer_time,1622754964
committer_tz,0


## `hvo5_project_summary.csv`

Semicolon-separated, one row per commit. `time` is the author timestamp in Unix seconds. The file is rewritten from the cache (not appended), so running this cell again does not duplicate rows.


In [7]:
full = pd.read_csv("df_commit_data.csv")
full = full.drop_duplicates(subset=["project", "commit"])
summary = full[["project", "commit", "author", "author_time", "message"]].copy()
summary["author_time"] = pd.to_numeric(summary["author_time"], errors="coerce")
missing_time = int(summary["author_time"].isna().sum())
if missing_time:
    print(f"WARNING: dropping {missing_time} rows with no author time")
    summary = summary.dropna(subset=["author_time"])
summary["author_time"] = summary["author_time"].astype("int64")
summary["message"] = summary["message"].fillna("")
summary = summary.rename(columns={
    "project": "project_wocid",
    "commit": "commit_sha1",
    "author_time": "time",
    "message": "commit message",
})
summary = summary.sort_values(["project_wocid", "time", "commit_sha1"]).reset_index(drop=True)
summary_path = f"{NETID}_project_summary.csv"
summary.to_csv(summary_path, sep=";", index=False)
print(f"wrote {summary_path}: {len(summary)} rows, {summary['project_wocid'].nunique()} projects")

# Round-trip check: the semicolon file must parse back to the same number of commits.
check = pd.read_csv(summary_path, sep=";")
assert list(check.columns) == ["project_wocid", "commit_sha1", "author", "time", "commit message"]
assert len(check) == len(summary)
print("columns:", list(check.columns))
check.groupby("project_wocid").size().rename("rows")


wrote hvo5_project_summary.csv: 53515 rows, 10 projects
columns: ['project_wocid', 'commit_sha1', 'author', 'time', 'commit message']


project_wocid
ablab_spades                      18459
cra-ros-pkg_robot_localization     2414
flatsurf_flatsurf                  2079
hmri-group_hmri-toolbox            1609
juliaastro_astrolib.jl              651
mano-b_microsec                     654
nocions_letswave6                   311
openspace_openspace               25493
photon-hdf5_phconvert               751
tdda_tdda                          1094
Name: rows, dtype: int64

## World of Code summary

For each project: number of commits, number of distinct author strings, and the earliest and latest author timestamps. The same person with two emails is counted twice, because WoC identifies authors by the name and email stored on the commit.


In [8]:
woc_stats = summary.groupby("project_wocid").agg(
    commits=("commit_sha1", "nunique"),
    authors=("author", "nunique"),
    min_time=("time", "min"),
    max_time=("time", "max"),
).reset_index()
woc_stats["min_time_utc"] = pd.to_datetime(woc_stats["min_time"], unit="s", utc=True).dt.strftime("%Y-%m-%d %H:%M:%S")
woc_stats["max_time_utc"] = pd.to_datetime(woc_stats["max_time"], unit="s", utc=True).dt.strftime("%Y-%m-%d %H:%M:%S")
woc_stats


,project_wocid,commits,authors,min_time,max_time,min_time_utc,max_time_utc
0,ablab_spades,18459,196,1295899946,1762802524,2011-01-24 20:12:26,2025-11-10 19:22:04
1,cra-ros-pkg_robot_localization,2414,240,1396281291,1759317757,2014-03-31 15:54:51,2025-10-01 11:22:37
2,flatsurf_flatsurf,2079,10,1547297444,1774299346,2019-01-12 12:50:44,2026-03-23 20:55:46
3,hmri-group_hmri-toolbox,1609,42,1413813574,1761738579,2014-10-20 13:59:34,2025-10-29 11:49:39
4,juliaastro_astrolib.jl,651,39,1457563164,1774899097,2016-03-09 22:39:24,2026-03-30 19:31:37
5,mano-b_microsec,654,8,1604123752,1760652530,2020-10-31 05:55:52,2025-10-16 22:08:50
6,nocions_letswave6,311,25,1427529750,1752135472,2015-03-28 08:02:30,2025-07-10 08:17:52
7,openspace_openspace,25493,254,1388750541,1776467917,2014-01-03 12:02:21,2026-04-17 23:18:37
8,photon-hdf5_phconvert,751,15,1422059137,1771428974,2015-01-24 00:25:37,2026-02-18 15:36:14
9,tdda_tdda,1094,16,1462464058,1757088306,2016-05-05 16:00:58,2025-09-05 16:05:06


## GitHub summary

Stars and forks come from the repository endpoint. The last commit date is the committer date of the latest commit on the default branch. The default-branch commit count and the contributor count are only for the validation check below. WoC is not expected to match them exactly.


In [9]:
def github_headers():
    headers = {
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        "User-Agent": "cs445-mp2-hvo5",
    }
    try:
        token = subprocess.check_output(
            ["gh", "auth", "token"], text=True, stderr=subprocess.DEVNULL
        ).strip()
    except Exception:
        token = ""
    if token:
        headers["Authorization"] = f"Bearer {token}"
    return headers


def last_page_count(response):
    """When per_page=1, the last page number is the item count."""
    link = response.headers.get("Link", "")
    match = re.search(r"[?&]page=(\d+)[^>]*>;\s*rel=\"last\"", link)
    if match:
        return int(match.group(1))
    if response.status_code == 200:
        body = response.json()
        if isinstance(body, list):
            return len(body)
    return None


session = requests.Session()
session.headers.update(github_headers())

gh_rows = []
for rec in assigned.itertuples(index=False):
    owner, repo = github_owner_repo(rec.GH)
    full_name = f"{owner}/{repo}"
    info = session.get(f"https://api.github.com/repos/{full_name}", timeout=60)
    info.raise_for_status()
    info_json = info.json()
    commits = session.get(
        f"https://api.github.com/repos/{full_name}/commits",
        params={"per_page": 1},
        timeout=60,
    )
    commits.raise_for_status()
    commit_json = commits.json()
    last_date = commit_json[0]["commit"]["committer"]["date"][:10] if commit_json else ""
    contributors = None
    for _ in range(3):
        contrib = session.get(
            f"https://api.github.com/repos/{full_name}/contributors",
            params={"per_page": 1, "anon": "true"},
            timeout=60,
        )
        if contrib.status_code == 202:
            time.sleep(3)
            continue
        contrib.raise_for_status()
        contributors = last_page_count(contrib)
        break
    gh_rows.append({
        "project_wocid": rec.WoC,
        "github": full_name,
        "stars": int(info_json["stargazers_count"]),
        "forks": int(info_json["forks_count"]),
        "last_commit_date": last_date,
        "gh_default_branch_commits": last_page_count(commits),
        "gh_contributors": contributors,
    })
    time.sleep(0.3)

gh_stats = pd.DataFrame(gh_rows)
gh_stats


,project_wocid,github,stars,forks,last_commit_date,gh_default_branch_commits,gh_contributors
0,juliaastro_astrolib.jl,JuliaAstro/AstroLib.jl,91,25,2026-07-08,317,24
1,cra-ros-pkg_robot_localization,cra-ros-pkg/robot_localization,1962,1039,2026-07-16,1074,81
2,hmri-group_hmri-toolbox,hMRI-group/hMRI-toolbox,86,46,2026-09-16,1463,24
3,tdda_tdda,tdda/tdda,311,29,2026-07-13,1419,12
4,openspace_openspace,OpenSpace/OpenSpace,1259,210,2026-09-26,12638,69
5,photon-hdf5_phconvert,Photon-HDF5/phconvert,17,14,2026-09-08,716,11
6,nocions_letswave6,NOCIONS/letswave6,50,32,2026-09-20,314,23
7,flatsurf_flatsurf,flatsurf/flatsurf,10,4,2025-12-17,1539,5
8,mano-b_microsec,MANO-B/MicroSEC,6,2,2025-10-16,653,2
9,ablab_spades,ablab/spades,961,165,2026-09-28,15354,140


## Validation

GitHub's commit count is the default branch only. WoC counts every commit it has stored for that project id, including other branches, and it stops around January 2025. Author counts also differ: WoC counts distinct `name <email>` strings, while GitHub's contributor list collapses identities and can omit some commits.


In [10]:
comparison = woc_stats.merge(gh_stats, on="project_wocid", how="outer")
comparison["commit_difference_woc_minus_github"] = (
    comparison["commits"] - comparison["gh_default_branch_commits"]
)
comparison["author_difference_woc_minus_github"] = (
    comparison["authors"] - comparison["gh_contributors"]
)
cols = [
    "project_wocid",
    "commits",
    "gh_default_branch_commits",
    "commit_difference_woc_minus_github",
    "authors",
    "gh_contributors",
    "author_difference_woc_minus_github",
    "max_time_utc",
    "last_commit_date",
]
comparison[cols]


,project_wocid,commits,gh_default_branch_commits,commit_difference_woc_minus_github,authors,gh_contributors,author_difference_woc_minus_github,max_time_utc,last_commit_date
0,ablab_spades,18459,15354,3105,196,140,56,2025-11-10 19:22:04,2026-09-28
1,cra-ros-pkg_robot_localization,2414,1074,1340,240,81,159,2025-10-01 11:22:37,2026-07-16
2,flatsurf_flatsurf,2079,1539,540,10,5,5,2026-03-23 20:55:46,2025-12-17
3,hmri-group_hmri-toolbox,1609,1463,146,42,24,18,2025-10-29 11:49:39,2026-09-16
4,juliaastro_astrolib.jl,651,317,334,39,24,15,2026-03-30 19:31:37,2026-07-08
5,mano-b_microsec,654,653,1,8,2,6,2025-10-16 22:08:50,2025-10-16
6,nocions_letswave6,311,314,-3,25,23,2,2025-07-10 08:17:52,2026-09-20
7,openspace_openspace,25493,12638,12855,254,69,185,2026-04-17 23:18:37,2026-09-26
8,photon-hdf5_phconvert,751,716,35,15,11,4,2026-02-18 15:36:14,2026-09-08
9,tdda_tdda,1094,1419,-325,16,12,4,2025-09-05 16:05:06,2026-07-13


## GitHub repository information

For each assigned project: number of stars, number of forks, and last commit date on the default branch (UTC), retrieved on 2026-09-28.

| Project (WoC id) | GitHub repository | Stars | Forks | Last commit date |
|---|---|---:|---:|---|
| juliaastro_astrolib.jl | JuliaAstro/AstroLib.jl | 91 | 25 | 2026-07-08 |
| cra-ros-pkg_robot_localization | cra-ros-pkg/robot_localization | 1962 | 1039 | 2026-07-16 |
| hmri-group_hmri-toolbox | hMRI-group/hMRI-toolbox | 86 | 46 | 2026-09-16 |
| tdda_tdda | tdda/tdda | 311 | 29 | 2026-07-13 |
| openspace_openspace | OpenSpace/OpenSpace | 1259 | 210 | 2026-09-26 |
| photon-hdf5_phconvert | Photon-HDF5/phconvert | 17 | 14 | 2026-09-08 |
| nocions_letswave6 | NOCIONS/letswave6 | 50 | 32 | 2026-09-20 |
| flatsurf_flatsurf | flatsurf/flatsurf | 10 | 4 | 2025-12-17 |
| mano-b_microsec | MANO-B/MicroSEC | 6 | 2 | 2025-10-16 |
| ablab_spades | ablab/spades | 961 | 165 | 2026-09-28 |


## World of Code information

For each assigned project: number of commits, number of authors (distinct `name <email>` strings), and minimum and maximum author time (Unix seconds, with the UTC date).

| Project (WoC id) | Commits | Authors | Min time | Max time |
|---|---:|---:|---|---|
| juliaastro_astrolib.jl | 651 | 39 | 1457563164 (2016-03-09) | 1774899097 (2026-03-30) |
| cra-ros-pkg_robot_localization | 2414 | 240 | 1396281291 (2014-03-31) | 1759317757 (2025-10-01) |
| hmri-group_hmri-toolbox | 1609 | 42 | 1413813574 (2014-10-20) | 1761738579 (2025-10-29) |
| tdda_tdda | 1094 | 16 | 1462464058 (2016-05-05) | 1757088306 (2025-09-05) |
| openspace_openspace | 25493 | 254 | 1388750541 (2014-01-03) | 1776467917 (2026-04-17) |
| photon-hdf5_phconvert | 751 | 15 | 1422059137 (2015-01-24) | 1771428974 (2026-02-18) |
| nocions_letswave6 | 311 | 25 | 1427529750 (2015-03-28) | 1752135472 (2025-07-10) |
| flatsurf_flatsurf | 2079 | 10 | 1547297444 (2019-01-12) | 1774299346 (2026-03-23) |
| mano-b_microsec | 654 | 8 | 1604123752 (2020-10-31) | 1760652530 (2025-10-16) |
| ablab_spades | 18459 | 196 | 1295899946 (2011-01-24) | 1762802524 (2025-11-10) |

WoC listed 53,598 commit SHA1s for these projects (`p2c`). Content was retrieved for 53,515. The other 83 (robot_localization 14, tdda 27, openspace 28, astrolib.jl 9, hmri-toolbox 4, flatsurf 1) are listed in `p2c` but missing from WoC's commit store: both `commit.tch` and `show_content('commit', ...)` return "Key not found". They are excluded from the counts above and from `hvo5_project_summary.csv`.


## Validation notes

- **Commit counts.** WoC counts are close to GitHub's default-branch counts for mano-b_microsec (654 vs 653), nocions_letswave6 (311 vs 314), photon-hdf5_phconvert (751 vs 716), and hmri-group_hmri-toolbox (1609 vs 1463). WoC is much higher for openspace_openspace (25,493 vs 12,638), cra-ros-pkg_robot_localization (2,414 vs 1,074), ablab_spades (18,459 vs 15,354), and juliaastro_astrolib.jl (651 vs 317). WoC keeps commits from every branch it has seen (robot_localization has a separate branch per ROS release, and astrolib.jl includes its `gh-pages` documentation builds), while GitHub's count covers the default branch only. tdda_tdda is the one project where WoC is lower (1,094 vs 1,419). About 325 commits are missing, 27 of them the unresolved SHA1s above, and the rest are commits made after WoC's last crawl of the project (WoC's latest tdda commit is 2025-09-05, and GitHub's is 2026-07-13).
- **Author counts.** WoC author counts are higher than GitHub contributor counts for every project, because WoC counts each distinct `name <email>` string, so one person using several emails or machines is counted more than once. GitHub merges identities by account and leaves out commits it cannot link to one.
- **Time range.** The version of WoC served by this API goes past the January 2025 cutoff mentioned in the assignment. Max times run from 2025-07 to 2026-04. Every project has GitHub commits after WoC's max time except mano-b_microsec, where the two dates are the same (2025-10-16), and flatsurf_flatsurf, where WoC's latest commit (2026-03-23) is newer than GitHub's default-branch date (2025-12-17), so it came from another branch.
